Bem commands [WIP]

lorenzo-desantis opened the pull request:
Converting bash scripts
Added mne/commands/mne_watershed_bem.py

agramfort commented on mne/commands/mne_watershed_bem.py:
```diff
@@ -0,0 +1,114 @@
+import mne
+
+
+def mne_watershed_bem(subject, subjects_dir, overwrite=False, volume='T1', atlas=False, gcaatlas=False, preflood=None):
```
subjects_dir should be None by default. If None use os.environ and raise error if not available.

also run pep8 on this file.

agramfort commented on mne/commands/mne_watershed_bem.py:
```diff
@@ -0,0 +1,114 @@
+        Subject name (SUBJECT environment variable)
+    subjects_dir : string
+        Directory containing subjects data (Freesurfer SUBJECTS_DIR environment variable)
+    overwrite
```
doctring is not OK

overwrite : bool

same remark below

agramfort commented on mne/commands/mne_watershed_bem.py:
```diff
@@ -0,0 +1,114 @@
+    os.environ['SUBJECTS_DIR'] = subjects_dir
+    env = os.environ.copy()
+
+    subject_dir = subjects_dir+'/'+subject
```
use os.path.join to join elements of path

agramfort commented on mne/commands/mne_watershed_bem.py:
```diff
@@ -0,0 +1,114 @@
+    """
+    os.environ['SUBJECT'] = subject
+    os.environ['SUBJECTS_DIR'] = subjects_dir
+    env = os.environ.copy()
```
check here that FREESURFER_HOME is set and raise error if not.

agramfort commented on mne/commands/mne_watershed_bem.py:
```diff
@@ -0,0 +1,114 @@
+    elif atlas:
+        cmd += ['-atlas']
+    if os.path.exists(T1_mgz):
+        cmd += ['-useSRAS', '-surf', ws_dir+'/'+subject, T1_mgz, ws_dir+'/ws']
```
same remark about joining path elements.

agramfort commented on mne/commands/mne_watershed_bem.py:
```diff
@@ -0,0 +1,114 @@
+    #
+    os.chdir(ws_dir)
+    if os.path.exists(T1_mgz):
+        surfaces = [subject+'_brain_surface', subject+'_inner_skull_surface',
```
pep8

agramfort commented on mne/commands/mne_watershed_bem.py:
```diff
@@ -0,0 +1,114 @@
+           '--id', '4', '--fif', subject+'-head.fif']
+    mne.utils.run_subprocess(cmd, env=env, stdout=sys.stdout)
+    print("Created %s/%s-head.fif" % (bem_dir, subject))
+    print("\nComplete.")
```
script should have a verbose argument and use mne logger. See how it's done in rest of the code base.

agramfort commented on mne/commands/mne_watershed_bem.py:
```diff
@@ -0,0 +1,114 @@
+
+is_main = (__name__ == '__main__')
+if is_main:
+    run()
```
you need to expose parameters to user with argparse. See other script in mne/commands

agramfort commented:
any progress here?

lorenzo-desantis commented:
I've added the usage of mne logger and argparser

agramfort commented:
can somebody give it a try?

in the mean time please travis which surely complain with pep8 violations

Eric89GXL commented on mne/commands/mne_watershed_bem.py:
```diff
@@ -0,0 +1,166 @@
+import os
+import shutil
+import mne
+from mne.utils import (verbose, logger)
```
might as well put `run_subprocess` up here, too

Eric89GXL commented on mne/commands/mne_watershed_bem.py:
```diff
@@ -0,0 +1,166 @@
+@verbose
+def mne_watershed_bem(subject=None, subjects_dir=None, overwrite=False,
+                      volume='T1', atlas=False, gcaatlas=False, preflood=None,
+                      verbose=None):
```
It probably makes sense to put this function in `mne.bem` instead of here in commands. And call it something more in line with the rest of `mne-python` like `make_watershed_bem` (the `mne` prefix is not meaningful because it will already be `mne.whatever`). That way it can be called directly from Python in addition to being called from a shell, which is the only way to get to it currently.

Eric89GXL commented on mne/commands/mne_watershed_bem.py:
```diff
@@ -0,0 +1,166 @@
+        try:
+            subjects_dir = os.environ['SUBJECTS_DIR']
+        except KeyError:
+            raise RuntimeError('SUBJECTS_DIR environment variable not defined')
```
We have a standard way of getting this through `get_subjects_dir`, see other uses throughout the code base. It should be something like `subjects_dir = get_subjects_dir(subjects_dir, raise_error=True)`

Eric89GXL commented on mne/commands/mne_watershed_bem.py:
```diff
@@ -0,0 +1,166 @@
+        try:
+            subject = os.environ['SUBJECT']
+        except KeyError:
+            raise RuntimeError('SUBJECT environment variable not defined')
```
Similarly I think you can get subject via `get_config` somehow, it checks the environment automatically

Eric89GXL commented on mne/commands/mne_watershed_bem.py:
```diff
@@ -0,0 +1,166 @@
+    bem_dir = os.path.join(subject_dir, 'bem')
+    ws_dir = os.path.join(subject_dir, 'bem', 'watershed')
+
+    if not os.path.exists(subject_dir):
```
For simplicity and consistency we use `import os.path as op` or `from os import path as op` and use `op.exists`, `op.join`, etc. -- it shortens a lot of lines.

Eric89GXL commented on mne/commands/mne_watershed_bem.py:
```diff
@@ -0,0 +1,166 @@
+import sys
+import os
+import shutil
+import mne
```
once you import `run_subprocess` directly, you won't need this import at all

lorenzo-desantis commented:
I corrected the watershed function and moved it to mne.bem as make_watershed_bem, while in commands there's still mne_watershed_bem.py which now uses make_watershed_bem

Eric89GXL commented:
We probably can't have a unit test because it would take forever, but it would still be good to have some way to test this manually at least. @lorenzo-desantis can you write a little script that you or a future dev could use to show that the script works? It could make a temporary directory (maybe using `mne.utils._TempDir`), calls it `subjects_dir`, copies the minimum necessary files from the `sample` subject in there, and then runs the algorithm?

I guess we could put the script in `examples/testing/test_bem_commands.py` or so. @agramfort WDYT?

Eric89GXL commented:
Also, if this is no longer WIP from your end, please update the title so that when scrolling through the open list of PRs we can triage appropriately.

Eric89GXL commented:
@lorenzo-desantis we are going to try to release v0.9 this week -- do you think you'll have time Monday or Tuesday to take a crack at wrapping this up?

lorenzo-desantis commented:
I'll try to do my best on Tuesday evening, I won't have time before.
Il 19/apr/2015 23:50, "Eric Larson" <notifications@github.com> ha scritto:

> @lorenzo-desantis <https://github.com/lorenzo-desantis> we are going to
> try to release v0.9 this week -- do you think you'll have time Monday or
> Tuesday to take a crack at wrapping this up?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1885#issuecomment-94317811>.
>

Eric89GXL commented:
Great, thanks for trying. If not, it's okay to put it into 0.10, too.

lorenzo-desantis commented:
Sorry but I didn't managed yesterday. 
Also, I have some problem extracting the testing dataset mne-testing-data-master.tar.gz after the download. Beside that, I've seen there are two dataset, mne.dataset.sample and mne.dataset.testing, which one should I use?

agramfort commented:
use sample for this.

agramfort commented:
any progress here?

@lorenzo-desantis we should see each other IRL and maybe @mainakjas @yousrabk and @teonlamont can give you some pointers.

lorenzo-desantis commented:
I had some problems with ArgvSetter, I haven't understood very clearly how
does it works.
About the meeting, would be fine this friday or next week?

2015-05-16 19:21 GMT+02:00 Alexandre Gramfort <notifications@github.com>:

> any progress here?
>
> @lorenzo-desantis <https://github.com/lorenzo-desantis> we should see
> each other IRL and maybe @mainakjas <https://github.com/mainakjas>
> @yousrabk <https://github.com/yousrabk> and @teonlamont
> <https://github.com/teonlamont> can give you some pointers.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1885#issuecomment-102657153>
> .
>

agramfort commented:
>
> I had some problems with ArgvSetter, I haven't understood very clearly how
> does it works.
> About the meeting, would be fine this friday or next week?
>

you can come on friday morning. If I have too little time I'll let you with
@yousrabk @teonlamont and @mainakjas . You will be in good hands :)

agramfort commented:
are you done here? travis is not happy...

let's keep the ball running here please.

agramfort commented:
are you done here?

lorenzo-desantis commented:
Do I move the test script with the other tests in
mne/commands/tests/test_commands.py or do I leave it in
examples/testing/test_bem_commands.py ?
Il 01/giu/2015 21:57, "Alexandre Gramfort" <notifications@github.com> ha
scritto:

> are you done here?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1885#issuecomment-107686713>
> .
>

agramfort commented:
move it with the others and mark it as slow

lorenzo-desantis commented:
Do I mark it also as @requires_freesurfer ?
Il 01/giu/2015 22:20, "Alexandre Gramfort" <notifications@github.com> ha
scritto:

> move it with the others and mark it as slow
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1885#issuecomment-107695318>
> .
>

agramfort commented:
Yes

lorenzo-desantis commented:
How do I mark that the script requires the sample dataset?

2015-06-02 0:01 GMT+02:00 Coveralls <notifications@github.com>:

> [image: Coverage Status] <https://coveralls.io/builds/2702462>
>
> Coverage decreased (-16.71%) to 73.5% when pulling *237f318
> <https://github.com/mne-tools/mne-python/commit/237f318a424cea3cb38957830b7b66a8ffff7fe2>
> on lorenzo-desantis:bem_commands* into *72b0077
> <https://github.com/mne-tools/mne-python/commit/72b00779ce46bedc12a4d143d19cceee6a6d6b5f>
> on mne-tools:master*.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1885#issuecomment-107726414>
> .
>

Eric89GXL commented:
@lorenzo-desantis right now no tests require the sample dataset. They all use the `testing` dataset.

You'll need to make a few additions to `mne/datasets/sample/sample.py` to have the `requires_sample_data` function like the `requires_testing_data` function in `mne/datasets/testing/_testing.py`. Once that is done, you should be able to do something like this:
```
from mne.datasets import sample

data_path = sample.data_path(download=False)

@sample.requires_sample_data
def test_thing():
    """Test something
    """
    pass
```

lorenzo-desantis commented:
The requires_testing_data function check for a preference key which is
'MNE_SKIP_TESTING_DATASET_TESTS', but there is not an analogous one for the
sample dataset, like 'MNE_SKIP_SAMPLE_DATASET_TESTS'. Do I skip the check
for this key or I leave it anyway?

2015-06-02 0:24 GMT+02:00 Eric Larson <notifications@github.com>:

> @lorenzo-desantis <https://github.com/lorenzo-desantis> right now no
> tests require the sample dataset. They all use the testing dataset.
>
> You'll need to make a few additions to mne/datasets/sample/sample.py to
> have the requires_sample_data function like the requires_testing_data
> function in mne/datasets/testing/_testing.py. Once that is done, you
> should be able to do something like this:
>
> from mne.datasets import sample
>
> data_path = sample.data_path(download=False)
>
> @sample.requires_sample_data
> def test_thing():
>     """Test something
>     """
>     pass
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1885#issuecomment-107734938>
> .
>

agramfort commented:
create one then?

lorenzo-desantis commented:
I found known_config_types which list the known keys values, I'll add MNE_SKIP_SAMPLE_DATASET_TESTS to the list.

agramfort commented:
good. Let us know when you're done so we can test.

agramfort commented:
travis is not happy. Can you look?

let's finish this asap and move on to other commands.

lorenzo-desantis commented:
Ok, there are no errors now!

agramfort commented:
can somebody give it a try?

in the mean time can you start doing the same for mne_flash_bem?

thanks heaps !

Eric89GXL commented:
I can try it tomorrow

agramfort commented:
if freesurfer is not setup it should raise a meaningful message. Not this:

```
Traceback (most recent call last):
  File "/Users/alex/work/src/mne-python/mne/commands/mne_watershed_bem.py", line 62, in <module>
    run()
  File "/Users/alex/work/src/mne-python/mne/commands/mne_watershed_bem.py", line 58, in run
    gcaatlas=gcaatlas, preflood=preflood, verbose=verbose)
  File "<string>", line 2, in make_watershed_bem
  File "/Users/alex/work/src/mne-python/mne/utils.py", line 541, in verbose
    return function(*args, **kwargs)
  File "/Users/alex/work/src/mne-python/mne/bem.py", line 372, in make_watershed_bem
    if not os.environ['FREESURFER_HOME']:
  File "/Users/alex/anaconda/lib/python2.7/UserDict.py", line 23, in __getitem__
    raise KeyError(key)
KeyError: 'FREESURFER_HOME'
```

agramfort commented on mne/bem.py:
```diff
@@ -332,3 +337,108 @@ def cost_fun(x, points):
+        raise RuntimeError('Could not find the MRI data')
+    if op.exists(ws_dir):
+        if not overwrite:
+            raise RuntimeError('%s already exists. Use the overwrite option to'
```
"overwrite option" -> "--overwrite option"

Eric89GXL commented on mne/commands/tests/test_commands.py:
```diff
@@ -177,4 +178,34 @@ def test_surf2bem():
+        mne_watershed_bem.run()
+
+    os.environ['SUBJECTS_DIR'] = orig_subject_dir
+    os.environ['SUBJECT'] = orig_subject
```
Can you see if it's possible to run this on the T1.mgz that is in the `testing` dataset? It is downsampled so it might fail, but if it passes it would be great because it might run faster. How long does this test take on your machine?

lorenzo-desantis replied:
The `test_watershed_bem` using the `sample` dataset it takes around 13 minutes.
Using the `testing` dataset it does run also, but the execution time is roughly the same (if not sligly longer).
So should I change the `test_watershed_bem` to use the `testing` dataset like the other tests or do I leave it with the `sample` dataset? If I use both the T1 folder from `sample` dataset and T1.mgz from the `testing` dataset, the watershed algorithm runs anyway on the latter.

agramfort replied:
I would use testing then mark the test as long

Eric89GXL replied:
If it takes that long we cannot run it in Travis, so we need to add another
category of slow-ness

Eric89GXL commented:
I got an error:
```
======================================================================
ERROR: Test mne watershed bem
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/larsoner/custombuilds/mne-python/mne/utils.py", line 616, in dec
    return function(*args, **kwargs)
  File "/home/larsoner/.local/lib/python2.7/site-packages/numpy/testing/decorators.py", line 146, in skipper_func
    return f(*args, **kwargs)
  File "/home/larsoner/custombuilds/mne-python/mne/commands/tests/test_commands.py", line 205, in test_watershed_bem
    mne_watershed_bem.run()
  File "/home/larsoner/custombuilds/mne-python/mne/commands/mne_watershed_bem.py", line 58, in run
    gcaatlas=gcaatlas, preflood=preflood, verbose=verbose)
  File "<string>", line 2, in make_watershed_bem
  File "/home/larsoner/custombuilds/mne-python/mne/utils.py", line 541, in verbose
    return function(*args, **kwargs)
  File "/home/larsoner/custombuilds/mne-python/mne/bem.py", line 426, in make_watershed_bem
    run_subprocess(cmd, env=env, stdout=sys.stdout)
  File "<string>", line 2, in run_subprocess
  File "/home/larsoner/custombuilds/mne-python/mne/utils.py", line 541, in verbose
    return function(*args, **kwargs)
  File "/home/larsoner/custombuilds/mne-python/mne/utils.py", line 783, in run_subprocess
    p = subprocess.Popen(command, *args, **kwargs)
  File "/usr/lib/python2.7/subprocess.py", line 702, in __init__
    errread, errwrite), to_close = self._get_handles(stdin, stdout, stderr)
  File "/usr/lib/python2.7/subprocess.py", line 1136, in _get_handles
    c2pwrite = stdout.fileno()
AttributeError: StringIO instance has no attribute 'fileno'
```
Since this error is in `subprocess`, you might need to use something other than `StringIO`, like actually create a temporary file (yuck) or something.

Eric89GXL commented on mne/commands/tests/test_commands.py:
```diff
@@ -177,4 +178,34 @@ def test_surf2bem():
+    """Test mne watershed bem"""
+    check_usage(mne_watershed_bem)
+    orig_subject_dir = os.getenv('SUBJECTS_DIR', None)
+    orig_subject = os.getenv('SUBJECT', None)
```
You get these values, but you never actually set `os.environ[...]`. So please add that bit.

Also, once you do that, please put a `try/except/finally` to set the `environ` back to normal no matter what errors get thrown.

Eric89GXL replied:
never mind about these comments, you should be able to omit this stuff entirely given my other comments above

lorenzo-desantis replied:
Is it actually necessary at all to get and set `os.environ[...]` here in test_commands.py?
In any case I pass SUBJECTS_DIR and SUBJECT as arguments and the checks to `os.environ` are done inside the function `make_watershed_bem` (where, following your comments below, I shouldn't modify at all `os.environ`).

Eric89GXL replied:
I don't think you should need to set the variables at all here

agramfort replied:
Freesurfer is Not on travis so we should be ok no?

Eric89GXL replied:
Yeah. But we should figure out some long term solution because this one
test will make tests take twice as long for most people...

Eric89GXL commented on mne/bem.py:
```diff
@@ -332,3 +337,108 @@ def cost_fun(x, points):
+    if not os.environ.get('FREESURFER_HOME'):
+        raise RuntimeError('FREESURFER_HOME environment variable not set')
+    if subject:
+        os.environ['SUBJECT'] = subject
```
you should just set `env` below so you don't modify `os.environ` at all

Eric89GXL commented on mne/bem.py:
```diff
@@ -332,3 +337,108 @@ def cost_fun(x, points):
+    else:
+        subject = get_config('SUBJECT', raise_error=True)
+    subjects_dir = get_subjects_dir(subjects_dir, raise_error=True)
+    os.environ['SUBJECTS_DIR'] = subjects_dir
```
same with this

Eric89GXL replied:
and you should ensure `subjects_dir` is not `None`, use `subjects_dir = get_subjects_dir(subjects_dir, raise_error=True)`

agramfort commented:
Are you sure your branch is up to date Eric?

Eric89GXL commented:
I just added the repo and tested 10 minutes ago...

Eric89GXL commented on mne/bem.py:
```diff
@@ -332,3 +337,108 @@ def cost_fun(x, points):
+# Create BEM surfaces
+
+@verbose
+def make_watershed_bem(subject=None, subjects_dir=None, overwrite=False,
```
I don't think we should allow `subject=None` here. It should be explicitly used. (Using the env var is not safe, most mne-pythoners don't use subject env vars as they aren't used anywhere else.)

lorenzo-desantis replied:
Here if the function `make_watershed_bem` is not given a `subject` argument, it checks for the presence of the `subject` key in the environment or in the mne configuration, and if not it raises an error through the function `get_config`.
Instead if the function `make_watershed_bem` is used through the script `mne_watershed_bem.py ` then it is prevented from running unless is explicilty given the `--subject [subject]` option.
Is it fine like that? Or should I modify the behaviour of the function?

Eric89GXL replied:
I don't think that the function in `mne/bem.py` should look for the `SUBJECT` env var, none of our other functions do this. As such, it's a little bit dangerous to do it actually. I'd prefer the command line tool not look for it, either, but it's understandable to have it look for the env var because that duplicates the functionality of the original shell script. Make sense?

lorenzo-desantis replied:
Ok, so the function `make_watershed_bem` will just raise an error if no `subject` is provided? Instead should it keep looking for the `SUBJECTS_DIR` var? In any case, the function must look for the `FREESURFER_HOME` env var.
Currently the script `mne_watershed_bem.py` do not check for env vars, but I don't think I'll have to add this check since the `subject` option is required to be set.

Eric89GXL replied:
Yeah, looking for `SUBJECTS_DIR` in the env is fine because we do that everywhere in the package. Same thing with FREESURFER_HOME, we look for that in the environment elsewhere as well. I'm glad `subject` is required, I think it's safer. Sounds good to me.

Eric89GXL commented on mne/commands/tests/test_commands.py:
```diff
@@ -177,4 +178,28 @@ def test_surf2bem():
 
 
+@slow_test
+@requires_freesurfer
```
Actually, I suspect Travis will fail on this because `FREESURFER_HOME` is set:

https://github.com/mne-tools/mne-python/blob/master/.travis.yml#L98

This is because we do run some tests with some of the basic freesurfer tools. We just don't have the correct command-line tools for this test, so it should fail. So maybe we do need to have an ultra-slow test switch, and not test those (only this one for now) by default. That will make it so Travis is happy, and so that users won't have their test time doubled. It does mean this function won't get tested often, but that is an okay tradeoff for me for now.

lorenzo-desantis replied:
Should I define a new ultra-slow marker for the test?

Eric89GXL replied:
I think we need to. @agramfort do you see a good alternative?

agramfort replied:
I agree on the proposition and the remarks from Eric.

Let me know when I shall review.

lorenzo-desantis replied:
I added an `@ultra_slow_test` decorator just like the existing `@slow_test` one. Is it fine?
How about the behaviour of Travis with respect to this new marker?

Eric89GXL commented:
You probably need to add `-a !ultra_slow_test` to the `nosetests` call on Travis, AppVeyor, and it the `Makefile` for the `test` (`make test`) target

Eric89GXL commented:
you can try it like `nosetsets --verbose -a !ultra_slow_test mne/commands/tests/` and it should skip the one you added

Eric89GXL commented:
if `-a !ultra_slow_test` isn't right, look at AppVeyor it already skips the slow ones somehow

lorenzo-desantis commented:
No problem with the testing now (the `@ultra_slow_test` are correctly skipped)

agramfort commented:
closing in favor of https://github.com/mne-tools/mne-python/pull/2194

I am taking over.

@lorenzo-desantis start doing the next command conversion.

thanks

agramfort commented:
check that I did not loose any of your commits

lorenzo-desantis commented:
Yes, no commit lost!
